bzerolib/channels/websocket/websocket.go

```
@@ -1,480 +1,255 @@
-		// Listener for any messages that need to be sent
-		ws.tmb.Go(func() error {
-			for ws.tmb.Alive() {
-				for ws.ready { // Might want a sync.Cond in the future if we're doing optimization
-					select {
-					case <-ws.tmb.Dying():
+		for {
+			select {
+			case <-ws.tmb.Dying():
+				ws.client.Close(fmt.Errorf("connection closing"))
+				return nil
+			case <-ws.client.Done():
+				ws.sendQueueReady = false
+				if autoReconnect {
+					logger.Infof("Broken connection to BastionZero, reconnecting...")
+
+					if err := ws.connect(connectionUrl, headers, params); err != nil {
+						logger.Errorf("failed to connect to BastionZero: %s", err)
 						return nil
-					case msg := <-ws.sendQueue:
-						ws.waitForAgentWebsocketReady()
-						ws.processOutput(msg)
 					}
-				}
-				time.Sleep(time.Second)
-			}
-			return nil
-		})
-
-		// Receive any messages in the websocket
-		for ws.tmb.Alive() {
-			for ws.ready { // Might want a sync.Cond in the future if we're doing optimization
-				select {
-				case <-ws.tmb.Dying():
+				} else {
+					logger.Infof("Connection with BastionZero closed and we're not retrying")
 					return nil
-				default:
-					if err := ws.receive(); err != nil {
-						ws.logger.Error(err)
-					}
+				}
+			case message := <-ws.client.Inbound():
+				if err := ws.receive(*message); err != nil {
+					ws.logger.Error(err)
 				}
 			}
 		}
-		return nil
 	})
 
 	return &ws, nil
 }
 
-func (w *Websocket) Close(reason error) {
-	w.logger.Infof("websocket closing because: %s", reason)
+func (w *Websocket) send() {
+	for {
+		select {
+		case <-w.tmb.Dead():
+			return
+		case message := <-w.sendQueue:
+			if !w.sendQueueReady && w.channelType == DaemonDataChannel {
+				w.waitForAgentWebsocketReady()
+			}
 
-	// close all of our existing datachannels
-	for _, channel := range w.channels {
-		channel.Close(reason)
+			if err := w.client.Send(*message); err != nil {
+				w.logger.Errorf("failed to send message: %s", err)
+			}
```

Dont we need to handle the case where w.sendQueueReady is false (not just for `DaemonDataChannel`)? It could be the case that the websocket disconnects and we are in the processing of reconnecting if autoReconnect is set to true. I think in these cases we would need to put the message back in the sendQueue?

---

okay very good point. I just realized in writing out an my entire logic here that it's based on having a bunch of sending channels which I changed and didn't notice this implication.

---

YES I got to do something a lot better and cooler 4b961e1

---

Ok i really like the changes you made with the `sendInterrupt` chan.

But i think there still may be race condition where the underlying websocket client closes and we fail to send  a message and never retry.

 The websocket closing causes the signalR done chan to finish https://github.com/bastionzero/bzero/blob/fd78d47eb1cbb55f131c6e9a3287ad018f2d7530/bzerolib/connection/messenger/signalr/signalr.go#L147

But before we get the interrupt signal its possible the send routine tries to send another message and this just fails silently and never gets resent. https://github.com/bastionzero/bzero/blob/274fbf93ffc1f6d45ea2407a3394b09504951cad/bzerolib/channels/websocket/websocket.go#L140-L142

---

alright. one more try. this time I removed the fancy and brought it back closer to what it was before. This should make more sense tho.